fn/mvar.go

```diff
@@ -0,0 +1,205 @@
+//
+// It supports two operations: take and put.
+//
+// The state transition rules are as follows:
+//  1. put while full blocks.
+//  2. put while empty sets.
+//  3. take while full resets.
+//  4. take while empty blocks.
+//  5. read while full nops.
+//  6. read while empty blocks.
+type MVar[A any] struct {
+	// value is the core state of the MVar
+	value *atomic.Pointer[A]
+
+	// mutex is a shared Mutex used by all of the Conds
+	mutex *sync.Mutex
+
+	// rCond is the Cond used to wake blocked Read operations.
+	rCond *sync.Cond
+
+	// tCond is the Cond used to wake blocked Take operations.
+	tCond *sync.Cond
+
+	// pCond is the Cond used to wake blocked Put operations.
+	pCond *sync.Cond
+
+	// reads is a WaitGroup that Put will wait for before signaling takers.
+	reads *sync.WaitGroup
+}
+
+// Zero initializes an MVar that has no values in it. In this state, TakeMVar
+// will block and PutMVar will immediately succeed.
+//
+// Zero : () -> MVar[A].
+func Zero[A any]() MVar[A] {
+	mutex := &sync.Mutex{}
+
+	return MVar[A]{
+		value: &atomic.Pointer[A]{},
+		mutex: mutex,
+		rCond: sync.NewCond(mutex),
+		tCond: sync.NewCond(mutex),
+		pCond: sync.NewCond(mutex),
+		reads: &sync.WaitGroup{},
+	}
+}
+
+// NewMVar initializes an MVar that has a value in it from the getgo. In this
+// state, TakeMVar will succeed immediately and PutMVar will block.
+//
+// NewMVar : A -> MVar[A].
+func NewMVar[A any](a A) MVar[A] {
+	z := Zero[A]()
+	z.value.Store(&a)
+	return z
+}
+
+// Take will wait for a value to be put into the MVar and then immediately
+// take it out.
+//
+// Take : MVar[A] -> A.
+func (m *MVar[A]) Take() A {
+	var val *A
+
+	// Here we use a Cond pattern and try to Swap nil into value. If we get
+	// a non-nil out it means we can actually finish the Take.
+	m.mutex.Lock()
+	for val = m.value.Swap(nil); val == nil; val = m.value.Swap(nil) {
+		m.tCond.Wait()
+	}
+	m.mutex.Unlock()
+
+	// now that we've successfully swapped a non-nil value out of the
+	// atomic, we can now wake any Put operations which would replace the
+	// value we just removed.
+	m.pCond.Signal()
+
+	return *val
+}
+
+// TryTake is the non-blocking version of TakeMVar, it will return an
+// None() Option if it would have blocked.
+//
+// TryTake : MVar[A] -> Option[A].
+func (m *MVar[A]) TryTake() Option[A] {
+	// If Swap is non-nil it means we can actually Take a value out.
+	if val := m.value.Swap(nil); val != nil {
+		// We just managed to Swap a nil in so we need to give blocked
+		// Puts the opportunity to proceed.
+		m.pCond.Signal()
+
+		return Some(*val)
+	}
+
+	return None[A]()
+}
+
+// Put will wait for a value to be made empty and will immediately replace it
+// with the argument.
+//
+// Put : (MVar[A], A) -> ().
+func (m *MVar[A]) Put(a A) {
+	// Here we use a Cond pattern and try to CompareAndSwap expecting value
+	// to be nil. If we successfully swap, it means we can actually finish
+	// the Put.
+	m.mutex.Lock()
```

Comment:
Need to double check, but I think API wise the lock _within_ the condition variable should be the one that's obtained.

Reply:
In this case they are the same.

Reply:
I have modified this to make it explicit so as not to spook anyone who is used to the pattern in the docs